python/eos/analysis.py

```diff
@@ -471,21 +472,30 @@ def sample_pmc(self, log_proposal, step_N=1000, steps=10, final_N=5000, rng=np.r
                 ))
             if last_perplexity < 0.05:
                 eos.warn("Last step's perplexity is very low. This could possibly be improved by running the markov chains that are used to form the initial PDF for a bit longer")
-            weights = sampler.weights[:][:, 0]
-            adjusted_weights = np.copy(weights)
-            # replace negative and nan weights by eps
-            adjusted_weights = np.where(
-                    np.logical_or(adjusted_weights <= 0, np.isnan(adjusted_weights)),
-                    eps, adjusted_weights)
+
+            # Use the samples of the last pmc_lookback steps to update the mixture
+            samples = sampler.samples[-pmc_lookback:]
+            weights = sampler.weights[-pmc_lookback:][:, 0]
             eos.info('Convergence diagnostics of all previous samples after sampling in step {}: perplexity = {}, ESS = {}'.format(
-                step, self._perplexity(adjusted_weights), self._ess(adjusted_weights)
+                step, self._perplexity(weights), self._ess(weights)
                 ))
-            pmc = pypmc.mix_adapt.pmc.PMC(samples, sampler.proposal, adjusted_weights, mincount=0, rb=True)
+
+            # Reevaluate the weights of the previous pmc_lookback steps
+            reevaluated_weights = pypmc.sampler.importance_sampling.combine_weights(
+                 samples.reshape(-1, step_N, len(self.varied_parameters)),
+                 weights.reshape(-1, step_N),
+                 proposals[-pmc_lookback:]
+                )[:][:,0]
+
+            pmc = pypmc.mix_adapt.pmc.PMC(samples, sampler.proposal, reevaluated_weights, mincount=0, rb=True)
             pmc.run(iterations=pmc_iterations, prune=0.0, rel_tol=pmc_rel_tol, abs_tol=pmc_abs_tol)
-            # Normalize the weights and remove components with a weight smaller than weight_threshold
             sampler.proposal = pmc.density
+            proposals.append(sampler.proposal)
```

Comment:
See above. Please check if making a copy of the proposal is necessary, and if no, record it in a comment.